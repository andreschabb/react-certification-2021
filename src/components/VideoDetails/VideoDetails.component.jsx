import React from 'react';
import { Icon } from 'semantic-ui-react';

import { YOUTUBE_BASE_URL } from '../../utils/constants';
import RelatedVideoItem from '../RelatedVideoItem';
import {
  BigCol,
  FavoritesButton,
  Row,
  SmallCol,
  TextVideoContainer,
  Title,
  VideoFrame,
  Description,
} from './VideoDetails.styles';

const VideoDetails = ({ videoId, description, title, videos }) => {
  const renderedVideos = videos
    .filter((video, index) => video.id.videoId !== videoId && index < 9)
    .map((video) => {
      return <RelatedVideoItem key={video.id.videoId} video={video} videos={videos} />;
    });

  return (
    <Row>
      <BigCol>
        <VideoFrame
          title="video-player"
          src={YOUTUBE_BASE_URL + videoId}
          allowFullScreen
        />
        <TextVideoContainer>
          <Title>{title}</Title>
          <FavoritesButton>
            Add To Favorites <Icon name="favorite" />
          </FavoritesButton>
        </TextVideoContainer>
        <hr />
        <Description>{description}</Description>
      </BigCol>
      <SmallCol>{renderedVideos}</SmallCol>
    </Row>
  );
};

export default VideoDetails;
