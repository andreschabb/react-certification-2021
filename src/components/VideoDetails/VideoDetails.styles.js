import styled from 'styled-components';

const redOrWhite = ({ isThemeLight }) => (isThemeLight ? 'red' : 'white');
const whiteOrRed = ({ isThemeLight }) => (isThemeLight ? 'white' : 'red');

export const Row = styled.div`
  &::after {
    content: '';
    display: table;
    clear: both;
  }
`;

export const BigCol = styled.div`
  float: left;
  width: 70%;
  padding: 10px;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);

  @media screen and (max-width: 950px) {
    width: 100%;
  }
`;

export const SmallCol = styled.div`
  float: left;
  width: 30%;
  padding: 10px;
  display: grid;
  grid-template-columns: 100%;

  @media screen and (max-width: 950px) {
    width: 100%;
  }
`;

export const VideoFrame = styled.iframe`
  width: 100%;
  height: 750px;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);

  @media screen and (max-width: 950px) {
    height: 500px;
  }
`;

export const TextVideoContainer = styled.div`
  padding: 2px;
  width: 100%;

  &::after {
    content: '';
    display: table;
    clear: both;
  }
`;

export const Title = styled.div`
  padding: 10px;
  width: 80%;
  float: left;
  text-transform: capitalize;
  font-weight: normal;
  text-align: left;
  font-size: 22px;

  @media screen and (max-width: 750px) {
    width: 100%;
  }
`;

export const FavoritesButton = styled.button`
  color: ${redOrWhite};
  background-color: ${whiteOrRed};
  padding: 10px;
  font-size: 16px;
  float: left;
  width: 20%;

  @media screen and (max-width: 750px) {
    width: 30%;
  }
  @media screen and (max-width: 550px) {
    width: 60%;
  }

  &:hover {
    border-color: ${redOrWhite};
    background-color: ${redOrWhite};
    color: ${whiteOrRed};
  }
`;

export const Description = styled.p`
  color: ${({ isThemeLight }) => (isThemeLight ? 'gray' : 'darkred')};
  font-size: 18px;
  text-align: justify;
`;
