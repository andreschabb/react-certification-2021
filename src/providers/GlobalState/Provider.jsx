import React, { createContext, useContext, useReducer } from 'react';

import reducer from './Reducer';

const initState = {
  isThemeLight: true,
};

const GlobalStateContext = createContext({
  isThemeLight: true,
});

function useGlobalState() {
  const context = useContext(GlobalStateContext);
  if (!context) {
    throw new Error(`Can't use "useTodos" without an TodoProvider!`);
  }
  return context;
}

function GloablStateProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, initState);

  return (
    <GlobalStateContext.Provider value={{ state, dispatch }}>
      {children}
    </GlobalStateContext.Provider>
  );
}

export { useGlobalState };

export default GloablStateProvider;
