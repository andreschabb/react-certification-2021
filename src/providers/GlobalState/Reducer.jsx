export default function reducer(state, action) {
  switch (action.type) {
    case 'CHANGE_THEME': {
      return {
        ...state,
        isThemeLight: action.payload,
      };
    }
    default:
      throw new Error('Unkown action');
  }
}
