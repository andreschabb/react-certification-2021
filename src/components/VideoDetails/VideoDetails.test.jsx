import React from 'react';
import { render, screen } from '@testing-library/react';

import VideoDetails from './VideoDetails.component';
import { useGlobalState } from '../../providers/GlobalState/Provider';

const mockedVideo = {
  id: { kind: 'youtube#video', videoId: 'nmXMgqjQzls' },
  snippet: {
    title: 'Video Tour | Welcome to Wizeline Guadalajara',
    description:
      'Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...',
    thumbnails: {
      default: {
        url: 'https://i.ytimg.com/vi/nmXMgqjQzls/default.jpg',
        width: 120,
        height: 90,
      },
    },
  },
};

const mockedVideos = [
  mockedVideo,
  {
    id: { kind: 'youtube#video', videoId: 'nmXMgqjQlhj' },
    snippet: {
      title: 'Title',
      description: 'Description',
      thumbnails: {
        default: {
          url: 'https://i.ytimg.com/vi/nmXMgqjQzls/default.jpg',
          width: 120,
          height: 90,
        },
      },
    },
  },
];

jest.mock('../../providers/GlobalState/Provider');

describe('VideoDetails', () => {
  beforeEach(() => {
    useGlobalState.mockImplementation(() => ({
      state: { isThemeLight: true },
    }));
  });

  it('should renders', () => {
    const { container } = render(
      <VideoDetails
        videoId={mockedVideo.id.videoId}
        description={mockedVideo.snippet.description}
        title={mockedVideo.snippet.title}
        videos={mockedVideos}
      />
    );

    expect(screen.getByText(mockedVideo.snippet.title)).toBeTruthy();
    expect(screen.getByText(mockedVideo.snippet.description)).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  it('should not render the detailed video as a related video item', () => {
    render(
      <VideoDetails
        videoId={mockedVideo.id.videoId}
        description={mockedVideo.snippet.description}
        title={mockedVideo.snippet.title}
        videos={mockedVideos}
      />
    );

    expect(screen.getAllByTestId('related-video-item').length).toBe(1);
    expect(screen.getByText(mockedVideos[1].snippet.title)).toBeTruthy();
  });
});
