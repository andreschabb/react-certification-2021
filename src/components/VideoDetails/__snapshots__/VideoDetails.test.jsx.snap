// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetails should renders 1`] = `
<div>
  <div
    class="sc-hKgILt bxZHCl"
  >
    <div
      class="sc-eCssSg bjWqBg"
    >
      <iframe
        allowfullscreen=""
        class="sc-gKsewC hODEnH"
        src="https://www.youtube.com/embed/nmXMgqjQzls"
        title="video-player"
      />
      <div
        class="sc-iBPRYJ dewWHJ"
      >
        <div
          class="sc-fubCfw hYiFMu"
        >
          Video Tour | Welcome to Wizeline Guadalajara
        </div>
      </div>
      <hr />
      <p
        class="sc-jrAGrp PtJMt"
      >
        Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...
      </p>
    </div>
    <div
      class="sc-jSgupP bEsdRU"
    >
      <div
        class="sc-bdfBwQ bMICcr"
        data-testid="related-video-item"
      >
        <img
          alt="Title"
          class="sc-gsTCUz dNFAFU"
          src="https://i.ytimg.com/vi/nmXMgqjQzls/default.jpg"
        />
        <p
          class="sc-dlfnbm ePZiwQ"
        >
          Title
        </p>
      </div>
    </div>
  </div>
</div>
`;
